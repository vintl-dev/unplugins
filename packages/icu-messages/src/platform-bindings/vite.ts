import type { UnpluginInstance, VitePlugin } from 'unplugin'
import { plugin as basePlugin, type Options } from '../plugin/index.ts'

export * from '../secondary-exports.ts'

export type PluginOptions = Options<VitePlugin>

export const icuMessages = (
  basePlugin as UnpluginInstance<PluginOptions, false>
).vite
