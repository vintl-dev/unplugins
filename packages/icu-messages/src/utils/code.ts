/**
 * Checks whether the provided code has singns of being already transformed.
 *
 * @param code The code to check.
 * @returns Whether the code has already been transformed.
 */
export function isTransformed(code_: string) {
  const code = code_.trim()

  for (const badToken of ['const', 'let', 'var', 'export', 'import']) {
    if (code.startsWith(badToken)) return true
  }

  return false
}
