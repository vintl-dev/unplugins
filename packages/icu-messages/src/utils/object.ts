/**
 * Accepts an object as input and checks whether the object is `undefined`,
 * `null`, or does not has own properties.
 *
 * @param value An object to check.
 * @returns Whether the object is `undefined`, `null`, or has no own properties.
 */
export function isEmptyObject(value?: Record<string, any>): boolean {
  if (value != null) {
    for (const key in value) {
      if (Object.hasOwn(value, key)) return false
    }
  }

  return true
}
