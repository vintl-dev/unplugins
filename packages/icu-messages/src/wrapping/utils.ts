/**
 * Checks whether the plugin wrapping is supported by the provided framework.
 *
 * @param framework Framework name to check in lower case.
 * @returns Whether the framework is one of the supported frameworks.
 */
export function isSupportedLoader(
  framework: string,
): framework is 'rollup' | 'vite' {
  return framework === 'rollup' || framework === 'vite'
}
