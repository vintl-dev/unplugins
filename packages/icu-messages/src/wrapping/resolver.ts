import type { WarningFunction } from '../shared/types.ts'
import { isEmptyObject } from '../utils/object.ts'
import type { WrappingFunctionsMap } from './options.ts'
import { wrapTransform } from './default-wrapper.ts'
import { PluginIneffectiveError } from './errors.ts'

/**
 * A class that holds a map of wrapping functions and exposes a function that
 * can be used to select a single function based on the plugin name.
 */
export class WrappingFunctionSelector<
  PluginType,
  InheritsDefaults extends boolean,
> {
  private readonly wrappers: WrappingFunctionsMap<PluginType>

  /**
   * Constructs a new wrapping functions selector.
   *
   * @param wrappers A map of wrapping functions.
   * @param inheritDefaults Whether to inherit the default wrapping functions.
   * @param onWarn A function that is called when there's a warning.
   */
  constructor(
    wrappers?: WrappingFunctionsMap<PluginType>,
    inheritDefaults?: InheritsDefaults,
    onWarn?: WarningFunction,
  ) {
    const useDefaults = inheritDefaults ?? true

    if (!useDefaults && isEmptyObject(wrappers)) {
      onWarn?.(
        new PluginIneffectiveError(
          'Your configuration does not make use of defaults and does not provide any other wrappers. This transform wrap plugin will be ineffective and probably could be removed.',
        ),
      )
    }

    this.wrappers = {
      ...(useDefaults &&
        ({
          json: wrapTransform,
          'vite:json': wrapTransform,
        } as WrappingFunctionsMap<PluginType>)),
      ...wrappers,
    }
  }

  /**
   * A function that takes in a plugin name and returns a wrapping function for
   * that plugin, if there's any. If there's none, returns `undefined`.
   *
   * It is bound to the object and can be destructured to be used on its own.
   *
   * @param pluginName The plugin name.
   * @returns The wrapping function, or `undefined`.
   */
  public readonly select = (pluginName: string) => {
    return this.wrappers[pluginName]
  }
}
