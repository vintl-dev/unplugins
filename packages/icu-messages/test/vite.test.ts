import type { OutputChunk } from 'rollup'
import { build } from 'vite'
import { describe, it, expect, assert } from 'vitest'
import { icuMessages } from '../dist/vite'
import { createResolver } from './utils/resolver'

const resolve = createResolver(import.meta.url)

describe('vite', () => {
  it('should generate bundle', async () => {
    const out = await build({
      plugins: [
        icuMessages({
          include: '**/*.messages.json',
          format: 'crowdin',
          pluginsWrapping: { use: true },
        }),
      ],
      build: {
        write: false,
        lib: {
          entry: resolve('fixtures/normal/input.mjs'),
          fileName: 'lib',
          formats: ['es'],
          name: 'TestOutput',
        },
      },
    })

    let entryChunk: OutputChunk | undefined

    if (Array.isArray(out)) {
      assert(out.length === 1, 'must have only one output')
      assert(out[0].output.length === 1, 'must produce only one chunk')
      entryChunk = out[0].output?.[0]
    } else {
      assert('output' in out, 'must not be a watcher')
      assert(out.output.length === 1, 'must produce only one chunk')
      entryChunk = out.output?.[0]
    }

    assert(entryChunk, 'must produce an entry chunk')

    expect(entryChunk.code).toMatchSnapshot()
  })
})
