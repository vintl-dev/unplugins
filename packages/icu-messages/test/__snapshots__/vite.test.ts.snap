// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vite > should generate bundle 1`] = `
"const e = [
  {
    type: 0,
    value: "Hello, "
  },
  {
    type: 1,
    value: "name"
  },
  {
    type: 0,
    value: "!"
  }
], t = {
  greeting: e
};
function a() {
  return t;
}
export {
  a as default
};
"
`;
